surfaces_tools/widgets/pdos.py

```diff
@@ -0,0 +1,727 @@
+import base64
+import copy
+import io
+import math
+import re
+
+import ipywidgets as ipw
+import matplotlib.pyplot as plt
+import numpy as np
+import traitlets as tl
+from aiida import common, orm
+from IPython.display import clear_output, display
+
+from ..helpers import HART_2_EV
+from ..utils import spm
+from . import stack
+
+
+def read_and_process_pdos_file(pdos_path):
+    try:
+        header = open(pdos_path).readline()
+    except TypeError:
+        header = pdos_path.readline()
+    try:  # noqa: TC101
+        kind = re.search(r"atomic kind.(\S+)", header).group(1)
+    except Exception:
+        kind = None
+    data = np.loadtxt(pdos_path)
+
+    # Determine fermi by counting the number of electrons and
+    # taking the middle of HOMO and LUMO.
+    n_el = int(np.round(np.sum(data[:, 2])))
+    if data[0, 2] > 1.5:
+        n_el = int(np.round(n_el / 2))
+    fermi = 0.5 * (data[n_el - 1, 1] + data[n_el, 1])
+
+    out_data = np.zeros((data.shape[0], 2))
+    out_data[:, 0] = (data[:, 1] - fermi) * HART_2_EV  # energy
+    out_data[:, 1] = np.sum(data[:, 3:], axis=1)  # "contracted pdos"
+    return out_data, kind
+
+
+def process_pdos_files(scf_calc, newversion=True):
+    if newversion:
+        retr_files = scf_calc.outputs.slab_retrieved.list_object_names()
+        retr_folder = scf_calc.outputs.slab_retrieved
+    else:
+        retr_files = scf_calc.outputs.retrieved.list_object_names()
+        retr_folder = scf_calc.outputs.retrieved
+
+    nspin = 1
+    for file in retr_files:
+        if file.endswith(".pdos") and "BETA" in file:
+            nspin = 2
+            break
+
+    dos = {
+        "mol": [None] * nspin,
+    }
+
+    for file in sorted(retr_files):
+        if file.endswith(".pdos"):
+            with retr_folder.open(file) as fhandle:
+                try:
```

Comment:
this type of try/except taht appears not only here had to be introduced with AiiDA 2.0. We still have many data requiring the older open procedure so for the moment we have to keep both